lumicks/pylake/kymotracker/detail/gaussian_mle.py

```diff
@@ -80,40 +80,45 @@ def normal_pdf_1d(x, center, sigma):
     return norm_factor * np.exp(-0.5 * ((x - center) / sigma) ** 2)
 
 
-def peak_expectation_1d(x, total_photons, center, sigma, background, pixel_size):
-    """Calculates the expectation value of a gaussian peak evaluated at x with baseline offset.
+def peak_expectation_1d(x, params, fixed_background, pixel_size, n_components):
+    """Calculates the expectation value of a sum of gaussian peaks evaluated at x with baseline
+    offset.
 
     Parameters
     ----------
     x : array-like
         Position data at which the function is to be evaluated.
-    total_photons : float
-        Total number of photons emitted by the imaged particle.
-    center : float
-        Peak center in um.
-    sigma : float
-        sigma parameter in um.
-    background : float
-        Background in photons per pixel.
+    params : array-like
+        Model parameters.
+        Model parameters should be provided as an array of parameters for each peak in the fit
+        followed by an optional offset parameter (in photons per pixel).
+        The peak parameters are: photon counts, center (in microns) and sigma (in microns).
+    fixed_background : float or None
+        Fixed background value in photons per pixel.
     pixel_size : float
         Pixel size in um.
     """
-    return total_photons * pixel_size * normal_pdf_1d(x, center, sigma) + background
+    signal = np.zeros(x.shape)
+    for total_photons, center, sigma in params[: n_components * 3].reshape(-1, 3):
+        signal += total_photons * pixel_size * normal_pdf_1d(x, center, sigma)
```

Comment:
`normal_pdf_1d` can be vectorized pretty straightforwardly. Broadcast `x` to a row shape `[1, x.size]` and each parameter to a column shape `[n_peaks, 1]`. The result will then be an array with the pdf for each peak as rows. I only know this off hand because I've done it before for HMM code where I had to optimize whereever I possibly could.

I think if we make this change we can skip the `for` loop here and in `peak_expectation_1d_derivatives`

Reply:
I just pushed a commit that vectorizes the code, but it's slower when I measure it.
```
Vectorized:     3.749 +- 0.0418
Non-vectorized: 3.403 +- 0.0570
```
I want to try reordering the parameters, to see if that has any sort of meaningful impact (right now I need a swapaxes and a reshape which I could possibly get around if I reorder the parameter values).

Reply:
Trying to do it over a different axes was worse.
```
Vectorized: 4.093789029121399 +- 0.1270268842186797
```

Reply:
And the last option, going over the same axis as the first but avoiding the `swapaxes` and `reshape`:
```
Vectorized: 3.523 +- 0.0817
```

Reply:
The last option is a bit faster when dealing with multiple peaks (previous tests were for single peak):
```python
Vectorized: 1.2794423580169678 +- 0.05959272465348172
Non-vectorized: 1.5789220094680787 +- 0.13292848948017963
```
Long story short, I'd go with vectorized option. And specifically with option 3, with the parameters grouped per type (avoiding the swap and reshape).